Add binary distribution cache manager take two

scottwittenburg opened the pull request:
This is a second attempt at #18359 as requested in #19147

scottwittenburg committed:
Move full_hash into spec schema

This key used to live in the buildcache_spec schema.  This change moves
it into the spec schema so it can exist with each spec anywhere that
schema is used (just buildcache_spec and database_index at the moment).

scottwittenburg committed:
Add cache management infrastructure to binary_distribution

This change adds a class which provides two levels of caching to the
module.

At one level, it caches buildcache index files from mirrors,
and uses a hash of the index contents to track whether the local caches
need to be updates.

Then the cached index files can be used to answer questions like "what
binary specs are available to me, based on my configured mirrors".  In
that case, the cached index files are used to populate the other level
of cache, the built spec cache.

The built spec cache in the cache manager replaces the previous global
"_cached_specs" variable with a cache that keeps track of which mirrors
contains which concrete specs.  However, if this cache does not contain
a target spec, the direct approach to fetching a spec.yaml from a mirror
is still supported.

One goal of this work is to support a new install option which will tell
spack to avoid installation from binary unless the full_hash of the binary
spec on te mirror mathes that of the local concrete spec.

scottwittenburg committed:
Provide --require-full-hash-match option to spack install command

This option defaults to false, preserving the previous behavior, which
was to install the first binary package found that matched the DAG hash
of the target spec.  If this option is provided, spack will install the
spec from source if it cannot find a binary spec on a remote mirror with
a full_hash that matches that of the local target spec.

scottwittenburg committed:
Fix fetch of spec.yaml files from buildcache (#19101)

Since those files currently exist in buildcaches (in S3 buckets) with
potentially different content types, we should be less restrictive in
what content types we accept when attempting to fetch them.  This PR
removes the content type constraint so any file with the matching
name will be found.

scottwittenburg committed:
Fix location in spec.yaml where we look for full_hash (#19132)

When we attempt to determine whether a remote spec (in a binary mirror)
is up-to-date or needs to be rebuilt, we compare the full_hash stored in
the remote spec.yaml file against the full_hash computed from the local
concrete spec.  Since the full_hash moved into the spec (and is no longer
at the top level of the spec.yaml), we need to look there for it.  This
oversight from #18359 was causing all specs to get rebuilt when the
full_hash wasn't fouhd at the expected location.

scottwittenburg committed:
Fix "buildcache list" command when mirror does not include index hash

If binary mirror was created before spack stored index hashes alongside
the index file, we should simply fetch the index, rather than complain
that the locally computed hash does not match the one (None) that does
not exist on the mirror.

scottwittenburg commented:
TLDR: 

Part of the suggestion in #19147 was to examine whether changing all the hashes was actually required for us to have the functionality provided by this PR.  ~~I think the answer is "yes"~~, and here are a few of my thoughts on the matter:

Details:

If we call `full_hash()` on a spec it forces it to get recomputed, so we can't do that to a concrete spec we have from a remote mirror, or it will always appear to match the locally computed one, which would defeat the purpose of all of this work.  So we need to rely on the `_full_hash` property that was last computed (which, for the purpose of this discussion, is when the binary package was created/pushed to the mirror).

We need the individual `spec.yaml` files on the mirror (which have a schema: `buildcache_spec.py`) to have the same notion of the buildcache state as the buildcache `index.json` (which also has a schema: `database_index.py`, also used by the database index, by the way).  Specifically, since the `spec.yaml` files record the full hash of the spec when it was built, so the `index.json` needs that information for all the specs too.  The `spec.yaml` files already store the full hash of the built spec, but not in a way that can be shared by the database.  Instead, this field is outside the spec, at the top level of the `spec.yaml`.  Since the `database_index` and `buildcache_spec` schemas both use the `spec` schema, moving the full hash into that schema seems like the sensible thing to do.

To achieve the goal of having the buildcache index and the individual `spec.yaml` files represent the same state, we need the `_full_hash` property of all specs to be serialized/deserialized by the database class when writing/reading buildcache index files.  ~~Since the database class uses the spec `to_node_dict()`/`from_node_dict()` methods to do that serialization/deserialization, I think the "node dict" representation of a spec needs to contain the full hash.~~

**Edit:** There is another set of methods on the spec, `to_dict` and 'from_dict`, which can be used instead of the previously chosen `to_node_dict` and `from_node_dict`.  By using `to_dict` and `from_dict` instead, we can serialize the full hash without letting it be used in the spec hash, thereby preventing the problem described in #19147.

scottwittenburg committed:
Remove full_hash from spec node_dict

This keeps the full hash living in the database index and in buildcache
entries (in the index as well as in individual spec.yaml files), but
removes it from the spec node dict, which influenced (broke) spec dag
hashing.

scottwittenburg commented:
@scheibelp Would you mind taking a look at this when you get a chance?  With that last commit ([fa537af](https://github.com/spack/spack/pull/19209/commits/fa537afc6a920cf455180f4d4f279691a2486567)), I can no longer reproduce #19147.  

Thanks @tgamblin for suggesting how to serialize/de-serialize the specs with their `full_hash`, and not have to change the hashes.

scottwittenburg commented on lib/spack/spack/database.py:
```diff
@@ -209,7 +209,12 @@ def to_dict(self, include_fields=default_install_record_fields):
-                rec_dict.update({'spec': self.spec.to_node_dict()})
+                spec_dict_list = self.spec.to_dict()['spec']
+                spec_dict = {}
+                for d in spec_dict_list:
```
This [comment](https://github.com/spack/spack/blob/89332361ba2a1c55ee3d0702ae517b10964535c9/lib/spack/spack/spec.py#L1621) made me wonder if we could skip iterating down the list returned by `to_dict()` here, and instead just access the first item?

scheibelp replied:
You could (since the traversal used to build the list is preorder, which guarantees that the root appears first). Although I think it's more straightforward to explicitly grab the spec by name here.

I should ask: is this needed for something? It looks like you are doing this to get the dictionary which includes the hash assigned by `Spec.to_dict`, but I don't see this dictionary being used anywhere (e.g. `InstallRecord.from_dict` discards the stored hash in favor of the one provided as a parameter to the function).

scottwittenburg replied:
The database is used to write both the traditional database index, as well as (more recently) the buildcache index.  The database `_write_to_file` method uses `InstallRecord.to_dict` on the specs it writes to the file.  So that part is needed and working for my purposes (of getting the `full_hash` written to both files).

> e.g. `InstallRecord.from_dict` discards the stored hash in favor of the one provided as a parameter to the function

I wanted to avoid changing how the database works for tracking installed specs, mostly because I didn't want to break anything.  But for reading back in a buildcache index, I'm using the database class (see [here](https://github.com/spack/spack/pull/19209/files#diff-d70c217a1f2343039a9033d745cf5a0517ac0fe42857dd8a5be6208bd4dc5b63R115-R123) for example), and it must get me the `full_hash` stored in the index rather than compute it anew.  Do we want similar functionality for tracking installed specs?  If so, then I should probably make sure the stored `full_hash` makes it from the index file back into the live `InstallRecord` instances.

scheibelp replied:
This looks correct and I see why it's needed, but I'm wondering if we can simplify this part of the code by adding a `Spec.node_dict_with_hash` function and e.g. replace 

```
        for s in self.traverse(order='pre', deptype=hash.deptype):
            node = s.to_node_dict(hash)
            node[s.name]['hash'] = s.dag_hash()
            if s.concrete:
                node[s.name]['full_hash'] = s.full_hash()
            if 'build' in hash.deptype:
                node[s.name]['build_hash'] = s.build_hash()
            node_list.append(node)
```

in `spec.py` with 

```
        for s in self.traverse(order='pre', deptype=hash.deptype):
            node_list.append(s.node_dict_with_hash(hash))
```

And `InstallRecord.to_dict` could use it as well

scottwittenburg commented on lib/spack/spack/database.py:
```diff
@@ -1112,6 +1117,7 @@ def _add(
             # the original hash of concrete specs.
             new_spec._mark_concrete()
             new_spec._hash = key
+            new_spec._full_hash = spec._full_hash
```
Should this be conditional, similar to `build_hash`, just below?  Sometimes we can get a concrete spec where the `full_hash` is `None`, we get that in the yaml.

scheibelp replied:
Apologies but I may be missing some context: I don't see `build_hash` mentioned in `database.py`

scottwittenburg replied:
Ah jeez, my bad.  That comment was intended for another section of the code, in `spec.py` [here](https://github.com/spack/spack/pull/19209/files#diff-9e28cdd6f9cd66e61101de6e9d22b1c33561959e4305c63147bbfbf7b22da936R1645).  Ugh, sorry for the confusion.

scottwittenburg commented:
Added a review with a couple questions...

scheibelp commented on lib/spack/spack/test/bindist.py:
```diff
@@ -508,3 +514,98 @@ def test_push_and_fetch_keys(mock_gnupghome):
+
+def test_spec_needs_rebuild(install_mockery_mutable_config, mock_packages,
+                            mock_fetch, monkeypatch, tmpdir):
+    """Make sure needs_rebuild properly comares remote full_hash
```
Typo: "comares" -> "compares"

scheibelp commented on lib/spack/spack/test/bindist.py:
```diff
@@ -508,3 +514,98 @@ def test_push_and_fetch_keys(mock_gnupghome):
+                         'install_dir_non_default_layout')
+def test_built_spec_cache(tmpdir,
+                          install_mockery):
+    """ Test what's the situation now """
```
Not sure what this comment is saying.

scheibelp commented on lib/spack/spack/binary_distribution.py:
```diff
@@ -37,23 +39,294 @@
+    1. cached buildcache index files (the index.json describing the built
+    specs available on a mirror)
+
+    2. a cache of all the conrcrete built specs available on all the
```
Typo: "conrcrete" -> "concrete"

scheibelp commented on lib/spack/spack/binary_distribution.py:
```diff
@@ -37,23 +39,294 @@
-'''
 
-BUILD_CACHE_INDEX_ENTRY_TEMPLATE = '  <li><a href="{path}">{path}</a></li>'
+class BinaryDistributionCacheManager(object):
```
(naming) IMO "Manager" doesn't add anything: I think "BinaryDistributionCache" is sufficient

scheibelp commented on lib/spack/spack/binary_distribution.py:
```diff
@@ -37,23 +39,294 @@
 
+
+#: default root, relative to the Spack install path
+default_manager_root = os.path.join(spack.paths.opt_path, 'spack')
```
(naming) I think `default_cache_root` would be more descriptive

Side note @carsonwoods this is another cache we will have to move to `~` in #11871

scheibelp commented on lib/spack/spack/binary_distribution.py:
```diff
@@ -536,30 +824,50 @@ def build_tarball(spec, outdir, force=False, rel=False, unsigned=False,
 
 
-def download_tarball(spec):
+def download_tarball(spec, url=None):
```
(naming) can this be called `preferred_url`?

scheibelp commented on lib/spack/spack/binary_distribution.py:
```diff
@@ -536,30 +824,50 @@ def build_tarball(spec, outdir, force=False, rel=False, unsigned=False,
+def download_tarball(spec, url=None):
     """
     Download binary tarball for given package into stage area
     Return True if successful
```
I think this part of the docstring is no longer accurate, can you update it?

scheibelp commented on lib/spack/spack/binary_distribution.py:
```diff
@@ -536,30 +824,50 @@ def build_tarball(spec, outdir, force=False, rel=False, unsigned=False,
             continue
 
+    # If we arrive here, something went wrong, maybe this should be an
+    # exception?
```
Could you elaborate? I think the issue is that if we are in this function, we have established that a binary cache claims to have a spec, but if we get to this point, we weren't actually able to download it. It may be worth printing a warning but as far as I can tell this was possible before this PR, so I don't think it should raise an error.

For example I think this could happen if there is an out-of-date index.json on the remote binary cache and the maintainer of that cache removed the file - which isn't an error on our end (I think the install should continue from source in that case).

scottwittenburg replied:
> I think the issue is that if we are in this function, we have established that a binary cache claims to have a spec, but if we get to this point, we weren't actually able to download it.

Yes, that's the idea, no matter how you got in here, we should have already verified that the spec exists on one of the mirrors.  I agree that installing from source when it's missing is the right thing to do.  I'll just print a warning though, as you suggest.

scheibelp commented on lib/spack/spack/binary_distribution.py:
```diff
@@ -843,117 +1151,94 @@ def extract_tarball(spec, filename, allow_root=False, unsigned=False,
 
-def get_specs():
+
+def get_spec(spec=None, force=False, full_hash_match=False):
```
(naming) can this be called `get_mirrors_for_spec`?

scheibelp commented on lib/spack/spack/binary_distribution.py:
```diff
@@ -843,117 +1151,94 @@ def extract_tarball(spec, filename, allow_root=False, unsigned=False,
+    def filter_candidates(possibles):
+        filtered_candidates = []
+        for candidate in possibles:
+            spec_full_hash = spec.full_hash()
```
A couple things:

* The full hash can be calculated outside of this function
* can the parameter `possibles` be renamed to `possibilities` or `candidates`?

scottwittenburg replied:
> * The full hash can be calculated outside of this function

Indeed, yikes!

scottwittenburg replied:
> * can the parameter `possibles` be renamed to `possibilities` or `candidates`?

I liked `candidates` better, but already had a variable of the same name in the outer method, so I went with `candidate_list` instead.

scheibelp commented on lib/spack/spack/installer.py:
```diff
@@ -345,13 +349,18 @@ def _try_install_from_binary_cache(pkg, explicit, unsigned=False):
-    binary_spec = spack.spec.Spec.from_dict(pkg.spec.to_dict())
-    binary_spec._mark_concrete()
-    if binary_spec not in specs:
+    matches = binary_distribution.get_spec(
```
(question) If I understand correctly, the updated logic ensures that `matches` includes all of the mirrors that store the binary cache (matching up to the full hash if specified). Why not pass this set to `_process_binary_cache_tarball` vs. passing a single member (in that case, you don't have to iterate through everything if the first match fails, you can just iterate through the set of matches)?

scottwittenburg replied:
Yeah, good suggestion.  I made a change to pass the list of matching urls to `download_tarball`, and try all of those first.  It still tries the remaining mirrors if all of those fail, even though I suppose that's not going to increase the likelihood of finding a tarball.  I wonder if `download_tarball` should check *either* the preferred mirrors *or* the configured mirrors, but not both?

scheibelp commented on lib/spack/spack/binary_distribution.py:
```diff
@@ -37,23 +39,294 @@
+                indexed_spec_dag_hash = indexed_spec.dag_hash()
+                if indexed_spec_dag_hash not in self._built_spec_cache:
+                    self._built_spec_cache[indexed_spec_dag_hash] = []
+                for entry in self._built_spec_cache[indexed_spec_dag_hash]:
```
This could use an explanatory `#`-style comment: it looks like you are trying to avoid registering the same `mirror_url` twice for a given spec. Since this function is only called once for each `mirror_url` I'm not sure how that would happen.

scottwittenburg replied:
I think my worry was that if `binary_distribution.get_specs` was called multiple times, I didn't want to append duplicate entries to the end of the list. But in that case, now that you point it out, I wonder if (when I encounter an entry with the same `mirror_url`) I should replace the `spec` key in that entry instead of just breaking.  When a subsequent call to `get_specs` results in a DAG hash match at a certain mirror, it seems possible it could actually be a new binary (with a different `full_hash`).

scheibelp commented on lib/spack/spack/binary_distribution.py:
```diff
@@ -37,23 +39,294 @@
+                                                    mirror_url)
+
+    def _associate_built_specs_with_mirror(self, cache_key, mirror_url):
+        self._init_local_index_cache()
```
I don't think this function needs to call `init_local_index_cache` each time since it appears to be a helper function for `regenerate_spec_cache` which already calls `init_local_index_cache`.

scheibelp commented on lib/spack/spack/binary_distribution.py:
```diff
@@ -37,23 +39,294 @@
+        self._index_contents_key = 'contents.json'
+        self._index_file_cache = None
+        self._local_index_cache = None
+        self._built_spec_cache = {}
```
For a given DAG hash, this contains a list of all binary caches which have a binary package for that DAG hash. I think this member could use a comment explaining that or an updated name like `self._mirrors_for_spec`.

scheibelp commented on lib/spack/spack/binary_distribution.py:
```diff
@@ -37,23 +39,294 @@
+            _, _, fs = web_util.read_from_url(hash_fetch_url)
+            fetched_hash = codecs.getreader('utf-8')(fs).read()
+        except (URLError, web_util.SpackWebError) as url_err:
+            tty.debug('Unable to read index hash {0}'.format(
```
(question) I'm assuming this is an expected pathway for older binary caches which don't have an `index.json.hash` file?

scottwittenburg replied:
That's correct, if we don't have it on the mirror, we print a debug message and keep going.  Later on, we check whether we actually got it before trying to compare it to the hash we compute locally on the fetched index.

scheibelp reviewed (changes requested):
I have a few questions and a large number of minor requests (mostly relating to typos or naming). I'll need to keep looking through this tomorrow.

scottwittenburg commented on lib/spack/spack/spec.py:
```diff
@@ -1642,6 +1642,7 @@ def to_dict(self, hash=ht.dag_hash):
         for s in self.traverse(order='pre', deptype=hash.deptype):
             node = s.to_node_dict(hash)
             node[s.name]['hash'] = s.dag_hash()
+            node[s.name]['full_hash'] = s._full_hash
```
Sorry @scheibelp this is where I should have put this question originally.  

Should this be conditional, similar to build_hash, just below? Sometimes we can get a concrete spec where the full_hash is None, we get that in the yaml.

scheibelp replied:
> Should this be conditional, similar to build_hash, just below?

It should not be: you always want the full hash available in the dictionary. Also this won't affect the other hashes. The build hash does not need to be computed conditionally either (although that's more of an aside).

One thing I don't understand yet: I don't see the `_full_hash` actually being calculated anywhere (which is why I'm not surprised when you say it is `None`) - I would have expected that you'd call `s.full_hash()` here (which computes the hash if it hasn't been already: `_full_hash` is a "caching" variable to avoid recomputing it).

scottwittenburg replied:
I see your point about why calling `full_hash()` might needed here, rather than using the cached property.  The thing I was most worried about was taking a concrete spec that was reconstituted from a `spec.yaml` on a mirror, and recomputing that hash locally, which would compute it based on the local `package.py`.  But that would be when reading from the dictionary index or `spec.yaml` (calling `from_dict`), rather than writing to them (and calling `to_dict`).

However, in my testing, it seems I'm always getting the right full hash written into the database index as well as into the buildcache index and mirror `spec.yaml` files, so I'm trying to figure out how that is working.  I'm wondering if it's just happening as a normal part of concretizing a spec.  I'm not sure yet, but I'll report back when I know what's going on here.

scottwittenburg replied:
So it's not during concretization that `full_hash()` is getting called, but during the install process as a part of trying to install from the binary cache.  It's also getting called a couple times from `database.py` in the `_failed_spec_path` method.  I don't understand what's going on behind the scenes there, maybe @tldahlgren can shed some light.  I tried to attach the log from a session installing `zlib` with a breakpoint set in the `full_hash()` method, but github continually failed to upload the file, so I'm pasting it below:

```
$ spack install zlib
> /data/scott/projects/spack/lib/spack/spack/spec.py(1466)full_hash()
-> return self._cached_hash(ht.full_hash, length)
(Pdb) p self.name
'zlib'
(Pdb) bt
  /data/scott/projects/spack/bin/spack(64)<module>()
-> sys.exit(spack.main.main())
  /data/scott/projects/spack/lib/spack/spack/main.py(762)main()
-> return _invoke_command(command, parser, args, unknown)
  /data/scott/projects/spack/lib/spack/spack/main.py(490)_invoke_command()
-> return_val = command(parser, args)
  /data/scott/projects/spack/lib/spack/spack/cmd/install.py(405)install()
-> install_spec(args, kwargs, abstract, concrete)
  /data/scott/projects/spack/lib/spack/spack/cmd/install.py(256)install_spec()
-> spec.package.do_install(**kwargs)
  /data/scott/projects/spack/lib/spack/spack/package.py(1602)do_install()
-> builder.install(**kwargs)
  /data/scott/projects/spack/lib/spack/spack/installer.py(1460)install()
-> self._init_queue(install_deps, install_package)
  /data/scott/projects/spack/lib/spack/spack/installer.py(1040)_init_queue()
-> spack.store.db.clear_failure(self.pkg.spec, force=True)
  /data/scott/projects/spack/lib/spack/spack/database.py(463)clear_failure()
-> if self.prefix_failure_marked(spec):
  /data/scott/projects/spack/lib/spack/spack/database.py(546)prefix_failure_marked()
-> return os.path.exists(self._failed_spec_path(spec))
  /data/scott/projects/spack/lib/spack/spack/database.py(416)_failed_spec_path()
-> '{0}-{1}'.format(spec.name, spec.full_hash()))
> /data/scott/projects/spack/lib/spack/spack/spec.py(1466)full_hash()
-> return self._cached_hash(ht.full_hash, length)
(Pdb) c
> /data/scott/projects/spack/lib/spack/spack/spec.py(1466)full_hash()
-> return self._cached_hash(ht.full_hash, length)
(Pdb) p self.name
'zlib'
(Pdb) bt
  /data/scott/projects/spack/bin/spack(64)<module>()
-> sys.exit(spack.main.main())
  /data/scott/projects/spack/lib/spack/spack/main.py(762)main()
-> return _invoke_command(command, parser, args, unknown)
  /data/scott/projects/spack/lib/spack/spack/main.py(490)_invoke_command()
-> return_val = command(parser, args)
  /data/scott/projects/spack/lib/spack/spack/cmd/install.py(405)install()
-> install_spec(args, kwargs, abstract, concrete)
  /data/scott/projects/spack/lib/spack/spack/cmd/install.py(256)install_spec()
-> spec.package.do_install(**kwargs)
  /data/scott/projects/spack/lib/spack/spack/package.py(1602)do_install()
-> builder.install(**kwargs)
  /data/scott/projects/spack/lib/spack/spack/installer.py(1501)install()
-> if pkg_id in self.failed or spack.store.db.prefix_failed(spec):
  /data/scott/projects/spack/lib/spack/spack/database.py(532)prefix_failed()
-> return self.prefix_failure_marked(spec)
  /data/scott/projects/spack/lib/spack/spack/database.py(546)prefix_failure_marked()
-> return os.path.exists(self._failed_spec_path(spec))
  /data/scott/projects/spack/lib/spack/spack/database.py(416)_failed_spec_path()
-> '{0}-{1}'.format(spec.name, spec.full_hash()))
> /data/scott/projects/spack/lib/spack/spack/spec.py(1466)full_hash()
-> return self._cached_hash(ht.full_hash, length)
(Pdb) c
==> Installing zlib
> /data/scott/projects/spack/lib/spack/spack/spec.py(1466)full_hash()
-> return self._cached_hash(ht.full_hash, length)
(Pdb) p self.name
'zlib'
(Pdb) bt
  /data/scott/projects/spack/bin/spack(64)<module>()
-> sys.exit(spack.main.main())
  /data/scott/projects/spack/lib/spack/spack/main.py(762)main()
-> return _invoke_command(command, parser, args, unknown)
  /data/scott/projects/spack/lib/spack/spack/main.py(490)_invoke_command()
-> return_val = command(parser, args)
  /data/scott/projects/spack/lib/spack/spack/cmd/install.py(405)install()
-> install_spec(args, kwargs, abstract, concrete)
  /data/scott/projects/spack/lib/spack/spack/cmd/install.py(256)install_spec()
-> spec.package.do_install(**kwargs)
  /data/scott/projects/spack/lib/spack/spack/package.py(1602)do_install()
-> builder.install(**kwargs)
  /data/scott/projects/spack/lib/spack/spack/installer.py(1591)install()
-> self._install_task(task, **kwargs)
  /data/scott/projects/spack/lib/spack/spack/installer.py(1081)_install_task()
-> full_hash_match):
  /data/scott/projects/spack/lib/spack/spack/installer.py(234)_install_from_cache()
-> pkg, explicit, unsigned=unsigned, full_hash_match=full_hash_match)
  /data/scott/projects/spack/lib/spack/spack/installer.py(355)_try_install_from_binary_cache()
-> pkg.spec, force=False, full_hash_match=full_hash_match)
  /data/scott/projects/spack/lib/spack/spack/binary_distribution.py(1227)get_mirrors_for_spec()
-> spec_full_hash = spec.full_hash()
> /data/scott/projects/spack/lib/spack/spack/spec.py(1466)full_hash()
-> return self._cached_hash(ht.full_hash, length)
(Pdb) c
==> No binary for zlib found: installing from source
==> Using cached archive: /data/scott/projects/spack/var/spack/cache/_source-cache/archive/c3/c3e5e9fdd5004dcb542feda5ee4f0ff0744628baf8ed2dd5d66f8ca1197cb1a1.tar.gz
==> zlib: Executing phase: 'install'
[+] /data/scott/projects/spack/opt/spack/linux-ubuntu18.04-haswell/gcc-7.4.0/zlib-1.2.11-gcrfowf2m3tywl4fh333mer3w2q6ytha
```

scottwittenburg replied:
So given that information, do you still think `full_hash()` needs to be called from the `spec.to_dict()` method?  I don't think it would hurt anything from my point of view (I don't think it would recompute/clobber it on a spec read in from yaml, which would be bad), and maybe it makes sense to do it just for sake of clarity when reading the code.

scheibelp replied:
> do you still think full_hash() needs to be called from the spec.to_dict()

Yes: we shouldn't depend on other parts to call `full_hash` and there is no extra cost because of the caching. When you update that, I'm curious if you still see the problem you mentioned in the first comment in this particular thread:

> Sometimes we can get a concrete spec where the full_hash is None, we get that in the yaml.

Regarding

> I don't think it would recompute/clobber it on a spec read in from yaml, which would be bad

I agree this would be a problem and we should not actually compute the full hash but instead pull it from the yaml.

scottwittenburg replied:
> Yes: we shouldn't depend on other parts to call `full_hash` and there is no extra cost because of the caching.

Sounds good, I'll call the method here instead of accessing the cached attribute.

> I'm curious if you still see the problem you mentioned in the first comment in this particular thread:
> 
> > Sometimes we can get a concrete spec where the full_hash is None, we get that in the yaml.

Sorry I wasn't clear about that.  It would happen when I would do something like the following in the spack python interpreter :

``` python
from spack.spec import Spec
s = Spec('zlib')
s.concretize()
with open('/some/path/spec.yaml', 'w') as fd:
    fd.write(s.to_dict())
```

scottwittenburg replied:
> > Yes: we shouldn't depend on other parts to call `full_hash` and there is no extra cost because of the caching.
> 
> Sounds good, I'll call the method here instead of accessing the cached attribute.

Just a heads up:  I made this change and it caused over a hundred tests to start failing, so I'm now looking into what's going on here.  I've verified that the commit before passes all those tests, and it's just the change from `... = s._full_hash` to `... = s.full_hash()` that causes the failures.

scottwittenburg replied:
There are a lot of failures of the form:

```
AttributeError: 'X' object has no attribute 'content_hash'
```

In stack traces like the following:

```
lib/spack/spack/test/database.py:90: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
lib/spack/spack/database.py:120: in converter
    return function(self, spec_like, *args, **kwargs)
lib/spack/spack/database.py:1140: in add
    self._add(spec, directory_layout, explicit=explicit)
lib/spack/llnl/util/lock.py:584: in __exit__
    if self._exit(release_fn):
lib/spack/llnl/util/lock.py:605: in _exit
    return self._lock.release_write(release_fn)
lib/spack/llnl/util/lock.py:461: in release_write
    result = release_fn()
lib/spack/llnl/util/lock.py:578: in release_fn
    return self._release_fn(type, value, traceback)
lib/spack/spack/database.py:983: in _write
    self._write_to_file(f)
lib/spack/spack/database.py:617: in _write_to_file
    installs = dict((k, v.to_dict(include_fields=self._record_fields))
lib/spack/spack/database.py:617: in <genexpr>
    installs = dict((k, v.to_dict(include_fields=self._record_fields))
lib/spack/spack/database.py:212: in to_dict
    spec_dict_list = self.spec.to_dict()['spec']
lib/spack/spack/spec.py:1645: in to_dict
    node[s.name]['full_hash'] = s.full_hash()
lib/spack/spack/spec.py:1464: in full_hash
    return self._cached_hash(ht.full_hash, length)
lib/spack/spack/spec.py:1431: in _cached_hash
    hash_string = self._spec_hash(hash)
lib/spack/spack/spec.py:1404: in _spec_hash
    node_dict = self.to_node_dict(hash=hash)
```

As I'm digging into it, it seems like maybe `MockPackageBase` is missing some properties required to handle calling `full_hash()` on a spec...

scottwittenburg replied:
Or maybe it would belong in the `MockPackageMultiRepo.add_package` method?

scheibelp replied:
Adding the following to `mock_package.MockPackageBase` fixes `spack test -k database` (I didn't run all tests):

```
    def content_hash(self):
        # Unlike real packages, MockPackage doesn't have a corresponding
        # package.py file; in that sense, the content_hash is always the same.
        return self.__class__.__name__
```

scottwittenburg replied:
Thanks @scheibelp I did something similar which fixed maybe 70% of the failures.  But I'll replace what I did with what you have here.

scottwittenburg replied:
I fixed several more of the failures by guarding the call to `full_hash()` inside `to_dict()` with a check for `spec.concrete()`.  If not concrete, I used the `_full_hash` property, but maybe `None` would be a better choice?

scheibelp replied:
This sounds good but IMO not setting `'full_hash'` when `not spec.concrete` would be clearest.

scottwittenburg replied:
Thanks for your help, I'm down to 22 failures.  Only a few of them fail when run in isolation, passing until they're run with (specific, but as yet unknown to me) other tests.  For example 5 of the `cmd/uninstall.py` tests fail when I run all the tests, but they all pass if I just run the `cmd/uninstall.py` tests.  Ditto for 7 of the `cmd/find.py` tests.

scheibelp replied:
Thanks for pursuing this - I think this is worthwhile:

* using `.full_hash()` will reduce fragility of `to_dict`
* the tests that are failing are likely fragile

scottwittenburg commented:
Thanks for your thoughtful review @scheibelp, I'm working my way through your suggestions, but won't finish until tomorrow.  Sounds like you may more feedback on this tomorrow anyway.

scottwittenburg committed:
Fix typos and make suggested renamings

scottwittenburg committed:
Pass download_tarball a list of preferred mirror urls instead of just one

This allows it to prioritize all mirrors that resulted in a spec match (up
to the full hash match, if that was required).  Other mirrors are still tried
if all preferred mirrors fail to result in a tarball download, even though
it's unlikely those will result in success.

scottwittenburg committed:
Replace cached spec entry in case full_hash has changed

In case we find we already have an entry for a mirror url under a
given DAG hash, we should replace the concrete spec we store, in
the case where it contains a different full_hash.

scottwittenburg committed:
Remove unused property

scottwittenburg committed:
Rename _built_spec_cache to _mirrors_for_spec

The new name better reflects the purpose/contents of this member,
which is now described by a comment.

scottwittenburg commented:
@scheibelp I just pushed my changes based on your review.  I left them as separate commits for now, to make it clear what I changed.  Modulo the discussion we're having about calling `full_hash()` vs using the `_full_hash` property, I think I've addressed all the concerns you've raised so far.

scottwittenburg committed:
Call full_hash() instead of accessing property when serializing to dict

scottwittenburg committed:
Fixes the following failing test:

    @pytest.mark.db
    def test_ambiguous_hash(self, mutable_database):
        x1 = Spec('a')
        x1._hash = 'xy'
        x1._concrete = True
        x2 = Spec('a')
        x2._hash = 'xx'
        x2._concrete = True
>       mutable_database.add(x1, spack.store.layout)

lib/spack/spack/test/spec_syntax.py:359:

lib/spack/spack/database.py:120: in converter
    return function(self, spec_like, *args, **kwargs)
lib/spack/spack/database.py:1140: in add
    self._add(spec, directory_layout, explicit=explicit)
lib/spack/llnl/util/lock.py:584: in __exit__
    if self._exit(release_fn):
lib/spack/llnl/util/lock.py:605: in _exit
    return self._lock.release_write(release_fn)
lib/spack/llnl/util/lock.py:461: in release_write
    result = release_fn()
lib/spack/llnl/util/lock.py:578: in release_fn
    return self._release_fn(type, value, traceback)
lib/spack/spack/database.py:983: in _write
    self._write_to_file(f)
lib/spack/spack/database.py:617: in _write_to_file
    installs = dict((k, v.to_dict(include_fields=self._record_fields))
lib/spack/spack/database.py:617: in <genexpr>
    installs = dict((k, v.to_dict(include_fields=self._record_fields))
lib/spack/spack/database.py:212: in to_dict
    spec_dict_list = self.spec.to_dict()['spec']
lib/spack/spack/spec.py:1646: in to_dict
    node[s.name]['full_hash'] = s.full_hash()
lib/spack/spack/spec.py:1464: in full_hash
    return self._cached_hash(ht.full_hash, length)
lib/spack/spack/spec.py:1431: in _cached_hash
    hash_string = self._spec_hash(hash)
lib/spack/spack/spec.py:1404: in _spec_hash
    node_dict = self.to_node_dict(hash=hash)
lib/spack/spack/spec.py:1562: in to_node_dict
    d['package_hash'] = self.package.content_hash()
lib/spack/spack/package.py:1423: in content_hash
    source_id = fs.for_package_version(self, self.version).source_id()

self = <spack.pkg.builtin.mock.a.A object at 0x10e78ff10>

    @property
    def version(self):
        if not self.spec.versions.concrete:
>           raise ValueError("Version requested for a package that"
                             " does not have a concrete version.")
E           ValueError: Version requested for a package that does not have a concrete version.

lib/spack/spack/package.py:812: ValueError

scottwittenburg committed:
Fix failing test:

The flake8 package both provides 'lapack' and violates a check
defined in fetch_strategy regarding the number of VCS url_attrs
allowed.

    def test_ci_generate_with_env(tmpdir, mutable_mock_env_path, env_deactivate,
                                  install_mockery, mock_packages):
        """Make sure we can get a .gitlab-ci.yml from an environment file
           which has the gitlab-ci, cdash, and mirrors sections."""
        filename = str(tmpdir.join('spack.yaml'))
        with open(filename, 'w') as f:
            f.write("""\
    spack:
      definitions:
        - bootstrap:
          - cmake@3.4.3
        - old-gcc-pkgs:
          - archive-files
          - callpath
          - hypre@0.2.15
      specs:
        - matrix:
          - [$old-gcc-pkgs]
      mirrors:
        some-mirror: https://my.fake.mirror
      gitlab-ci:
        bootstrap:
          - name: bootstrap
            compiler-agnostic: true
        mappings:
          - match:
              - arch=test-debian6-x86_64
            runner-attributes:
              tags:
                - donotcare
              image: donotcare
        final-stage-rebuild-index:
          image: donotcare
          tags: [donotcare]
      cdash:
        build-group: Not important
        url: https://my.fake.cdash
        project: Not used
        site: Nothing
    """)
        with tmpdir.as_cwd():
            env_cmd('create', 'test', './spack.yaml')
            outputfile = str(tmpdir.join('.gitlab-ci.yml'))

            with ev.read('test'):
>               ci_cmd('generate', '--output-file', outputfile)

lib/spack/spack/test/cmd/ci.py:153:

lib/spack/spack/main.py:545: in __call__
    self.returncode = _invoke_command(
lib/spack/spack/main.py:490: in _invoke_command
    return_val = command(parser, args)
lib/spack/spack/cmd/ci.py:387: in ci
    args.func(args)
lib/spack/spack/cmd/ci.py:88: in ci_generate
    spack_ci.generate_gitlab_ci_yaml(
lib/spack/spack/ci.py:660: in generate_gitlab_ci_yaml
    'SPACK_ROOT_SPEC': format_root_spec(
lib/spack/spack/ci.py:195: in format_root_spec
    spec_yaml = spec.to_yaml(hash=ht.build_hash).encode('utf-8')
lib/spack/spack/spec.py:1680: in to_yaml
    self.to_dict(hash), stream=stream, default_flow_style=False)
lib/spack/spack/spec.py:1646: in to_dict
    node[s.name]['full_hash'] = s.full_hash()
lib/spack/spack/spec.py:1464: in full_hash
    return self._cached_hash(ht.full_hash, length)
lib/spack/spack/spec.py:1431: in _cached_hash
    hash_string = self._spec_hash(hash)
lib/spack/spack/spec.py:1404: in _spec_hash
    node_dict = self.to_node_dict(hash=hash)
lib/spack/spack/spec.py:1566: in to_node_dict
    d['dependencies'] = syaml.syaml_dict([
lib/spack/spack/spec.py:1569: in <listcomp>
    ('hash', dspec.spec._cached_hash(hash)),
lib/spack/spack/spec.py:1431: in _cached_hash
    hash_string = self._spec_hash(hash)
lib/spack/spack/spec.py:1404: in _spec_hash
    node_dict = self.to_node_dict(hash=hash)
lib/spack/spack/spec.py:1562: in to_node_dict
    d['package_hash'] = self.package.content_hash()
lib/spack/spack/package.py:1423: in content_hash
    source_id = fs.for_package_version(self, self.version).source_id()
lib/spack/spack/fetch_strategy.py:1305: in for_package_version
    check_pkg_attributes(pkg)

pkg = <spack.pkg.builtin.mock.flake8.Flake8 object at 0x108c25e20>

    def check_pkg_attributes(pkg):
        """Find ambiguous top-level fetch attributes in a package.

        Currently this only ensures that two or more VCS fetch strategies are
        not specified at once.
        """
        # a single package cannot have URL attributes for multiple VCS fetch
        # strategies *unless* they are the same attribute.
        conflicts = set([s.url_attr for s in all_strategies
                         if hasattr(pkg, s.url_attr)])

        # URL isn't a VCS fetch method. We can use it with a VCS method.
        conflicts -= set(['url'])

        if len(conflicts) > 1:
            print('more than one conflict for {0}'.format(pkg.name))
            print(conflicts)
>           raise FetcherConflict(
                'Package %s cannot specify %s together. Pick at most one.'
                % (pkg.name, comma_and(quote(conflicts))))
E           spack.fetch_strategy.FetcherConflict: Package flake8 cannot specify 'svn', 'hg', and 'git' together. Pick at most one.

lib/spack/spack/fetch_strategy.py:1239: FetcherConflict

scheibelp commented on lib/spack/spack/binary_distribution.py:
```diff
@@ -37,23 +39,307 @@
+                hash_fetch_url), url_err, 1)
+
+        # If we were expecting some hash and found that we got it, we're done
+        if expect_hash and fetched_hash == expect_hash:
```
If we call this with the default `fetch_hash`, and we fail to retrieve the remote hash file, it looks like we would incorrectly return here (since both `fetch_hash` and `expect_hash` would be `None`).

scheibelp commented on lib/spack/spack/binary_distribution.py:
```diff
@@ -37,23 +39,307 @@
+    1. cached buildcache index files (the index.json describing the built
+    specs available on a mirror)
+
+    2. a cache of all the concrete built specs available on all the
```
Is this a cache or more of a reverse mapping of `1`? I think the following description would be clearer:

```
The BinaryDistributionCache stores "index.json" files from binary caches and can report

1. All specs that are stored in some binary cache
2. For any such spec, the list of binary caches where it can be retrieved
```

scottwittenburg replied:
It's not strictly a reverse mapping of `1`, though I suppose that's one way to think of it.  If you don't call `spack buildcache list`, you won't have cache of `index.json` files from remote mirrors, but if you `spack install` some package, and there is a cache hit, you would have that concrete spec in your binary spec cache.

Should this description go into that level of detail?

scheibelp replied:
I'm not necessarily looking for more detail - primarily I want a more user-oriented description, e.g. "user X can get information Y" from this object. My suggested rewording wasn't as specific as suggesting a reverse mapping, but was it also wrong?

scheibelp replied:
The updated docstring is useful but I think the intro could summarize at a higher level:

```
"""
The BinaryCacheIndex tracks what specs are available on
(usually remote) binary caches

This index is "best effort", in the sense that whenever we don't find what...
```

scheibelp commented on lib/spack/spack/installer.py:
```diff
@@ -302,7 +303,8 @@ def _process_external_package(pkg, explicit):
 
-def _process_binary_cache_tarball(pkg, binary_spec, explicit, unsigned):
+def _process_binary_cache_tarball(pkg, binary_spec, explicit, unsigned,
+                                  preferred_mirrors=[]):
```
I recommend setting `preferred_mirrors=None` as the default and if it needs to be a list instantiate it below like

```
binary_distribution.download_tarball(
         binary_spec, preferred_mirrors=(preferred_mirrors or list()))
```

(this isn't necessarily wrong, but should only be done if necessary, see discussions on default mutable arguments in python for more).

scheibelp commented on lib/spack/spack/test/spec_syntax.py:
```diff
@@ -349,13 +349,16 @@ def test_multiple_specs_with_hash(self, database):
         x2._hash = 'xx'
         x2._concrete = True
+
+        monkeypatch.setattr(spack.package.PackageBase, 'version', '1.0')
```
I think it would be more straightforward to concretize x1/x2 and then explicitly set their hashes vs. setting properties on the associated `Package` object, like so:

```
        x1 = Spec('a')
        x1.concretize()
        x1._hash = 'xy'
        x2 = Spec('a')
        x2.concretize()
        x2._hash = 'xx'

        mutable_database.add(x1, spack.store.layout)
        mutable_database.add(x2, spack.store.layout)
```

scheibelp commented on lib/spack/spack/test/cmd/ci.py:
```diff
@@ -119,7 +119,7 @@ def test_ci_generate_with_env(tmpdir, mutable_mock_env_path, env_deactivate,
       - archive-files
       - callpath
-      - hypre@0.2.15
+      - hypre@0.2.15 ^openblas-with-lapack
```
Could you add a comment here explaining this:

> specify `^openblas-with-lapack` to ensure that `builtin.mock repo flake8` package (which can also provide lapack) is not chosen, as it violates a package-level check which requires exactly one fetch strategy (this is apparently not an issue for other tests that use it).

scheibelp commented on lib/spack/spack/binary_distribution.py:
```diff
@@ -545,30 +846,50 @@ def build_tarball(spec, outdir, force=False, rel=False, unsigned=False,
 
 
-def download_tarball(spec):
+def download_tarball(spec, preferred_mirrors=[]):
```
Default initializer issue (use `preferred_mirrors=None`).

scheibelp commented on lib/spack/spack/test/installer.py:
```diff
@@ -177,7 +177,7 @@ def test_process_binary_cache_tarball_none(install_mockery, monkeypatch,
 def test_process_binary_cache_tarball_tar(install_mockery, monkeypatch, capfd):
     """Tests of _process_binary_cache_tarball with a tar file."""
-    def _spec(spec):
+    def _spec(spec, preferred_mirrors=[]):
```
Default initializer issue (use `preferred_mirrors=None`).

scheibelp reviewed (changes requested):
I have a few more requests. Let me know if they seem off track.

scottwittenburg committed:
A few more review changes

scheibelp commented on lib/spack/spack/binary_distribution.py:
```diff
@@ -37,23 +39,307 @@
+                        'spec': new_entry['spec'],
+                    }
+
+    def update_local_index_cache(self):
```
I'm thinking about the use of the word "cache" here in general and how to distinguish between:

* The binary cache, i.e. the (often) remotely-hosted store of already-built Spack packages
* The index, i.e. this object

This index *does* cache files (i.e. the remote index files), so perhaps the object should be called `BinaryDistributionIndexCache` (and the docstring should mention that distinction). If you don't think that's clearer, then I'm wondering if the word "cache" should be dropped from these function names.

scottwittenburg replied:
Because it caches both index files and concrete specs, I had initially named it (maybe somewhat vaguely) `BinaryDistributionCacheManager`.  It does manage caches of two related kinds of things.  But the word "cache" is getting pretty heavily overloaded, so I agree it makes sense to try to clear things up to reduce potential confusion.

I put `index_cache` in that function name because it's updating the locally stored binary cache index files, as opposed to the locally stored cache of concrete specs.

I'm not vehemently opposed to either of your renaming suggestions here, but neither one feels quite right to me.  While it does cache index files, it doesn't *only* cache them, so I'm not sure I like making the class name more specific in that way.  And regarding dropping `index` from the method name `update_local_index_cache`, to me that makes the purpose of the method more ambiguous.  That method doesn't update the cache of concrete specs, only the cache of index files.

I'm worried there must be something more "overarchingly weird" in this change, making it difficult to nail down exactly what's confusing about it.  Somehow I think we need to figure out what that weirdness is.  Maybe we can sync up in person at some point to try to get to the bottom of it?

scheibelp replied:
I think we resolved most of this, but I think the name `update` would be suitable (e.g. `update_local` isn't necessary because there is no `update_remote` option for this object).

scottwittenburg committed:
Improve method documentation, renaming, and a couple fixes

This change ensures that on a "spack install", we first initialize
the cache objects from any cached-on-disk buildcache index.json
files we may have retrieved previously.  It also improves some of
the method and class documentation, as well as renames some methods
to better reflect what they do,  Additionally, when constructing
the internal concrete spec cache (_mirrors_for_spec) from the
locally cached index.json files, it makes sure to only perform
the work for that association once, using the hash of the index.json
as a key to guard against doing it multiple times.

scottwittenburg commented:
@scheibelp I've made the changes you suggested here as well as some more we discussed earlier today, this is ready for another look when you have a chance.  Thanks!

scheibelp commented on lib/spack/spack/test/spec_syntax.py:
```diff
@@ -349,13 +349,14 @@ def test_multiple_specs_with_hash(self, database):
 
     @pytest.mark.db
-    def test_ambiguous_hash(self, mutable_database):
+    def test_ambiguous_hash(self, mutable_database, monkeypatch):
```
Thanks for updating this. I don't think the `monkeypatch` fixture is needed anymore.

scheibelp commented on lib/spack/spack/binary_distribution.py:
```diff
@@ -37,23 +39,349 @@
+
+
+#: Singleton cache_manager instance
+cache_manager = llnl.util.lang.Singleton(_cache_manager)
```
(naming) Could this be called `binary_index` and `_cache_manager` be renamed to `_binary_index`?

scheibelp commented on lib/spack/spack/binary_distribution.py:
```diff
@@ -37,23 +39,349 @@
+        self._specs_already_associated = {}
+
+        # _mirrors_for_spec is a dictionary mapping DAG hashes to lists of
+        # entries indicating mirrors where that concrete spec can be found:
```
I think it's worth mentioning that each entry consists of 

* the mirror where the spec is
* the spec itself (including the full hash, since the dag hash may match but we want to use the updated source if available).

scheibelp commented on lib/spack/spack/binary_distribution.py:
```diff
@@ -37,23 +39,349 @@
+def _cache_manager():
+    """Get the singleton store instance."""
+    cache_root = spack.config.get(
+        'config:binary_distribution_cache_root', default_cache_root)
```
(naming) Could this be called `config:binary_index_root` and `default_binary_index_root`?

scheibelp commented on lib/spack/spack/binary_distribution.py:
```diff
@@ -37,23 +39,349 @@
-'''
 
-BUILD_CACHE_INDEX_ENTRY_TEMPLATE = '  <li><a href="{path}">{path}</a></li>'
+class BinaryDistributionCache(object):
```
(naming) Can this be renamed to `BinaryCacheIndex`?

scheibelp commented on lib/spack/spack/binary_distribution.py:
```diff
@@ -37,23 +39,349 @@
+                        cache_key) as cache_file:
+                    self._local_index_cache = json.load(cache_file)
+
+    def clear_spec_cache(self):
```
A few things:

* Should `_local_index_cache` also be cleared here?
* Can you mention this is for testing?
* Should this delete the `_index_file_cache` (setting it to `None` won't remove the underlying files)?

(naming) I suggest the function name `clear`.

scheibelp commented on lib/spack/spack/binary_distribution.py:
```diff
@@ -37,23 +39,349 @@
+    def __init__(self, cache_root=None):
+        self._cache_root = cache_root or default_cache_root
+        self._index_cache_root = os.path.join(self._cache_root, 'indices')
+        self._index_contents_key = 'contents.json'
```
Each of these attributes could use a `#`-style comment:

* `_local_index_cache`: stores a map of mirror URL to index hash file
* `_index_file_cache`: stores copies of remote binary cache indices as well as a serialized `_local_index_cache`
* `_index_contents_key`: the key associated with the serialized `_local_index_cache` in the `_index_file_cache`

scheibelp commented on lib/spack/spack/binary_distribution.py:
```diff
@@ -37,23 +39,349 @@
+            url_hash[:10], locally_computed_hash[:10])
+        self._index_file_cache.init_entry(cache_key)
+        with self._index_file_cache.write_transaction(cache_key) as (old, new):
+            new.write(index_object_str)
```
I see the new cache_key for the `mirror_url` being written, but is the old one being deleted? Should that be done here before overwriting `mirror_url`? This isn't a major problem but could end up cluttering the file cache with many outdated `index.json.hash` files.

scheibelp commented on lib/spack/spack/binary_distribution.py:
```diff
@@ -37,23 +39,349 @@
+                        }
+                    ]
+        """
+        self.regenerate_spec_cache()
```
This looks like what is needed to lazily populate the cache. However, this may end up doing a significant amount of work for each call. I think it would make sense to add a function called `_read()` like:

```
def _read(self):
  if self._mirrors_for_spec:
    return
  self.regenerate_spec_cache()
```

and call that here.

scottwittenburg replied:
In my last round of changes, I added some code to protect against `regenerate_spec_cache()` doing a lot of work if it didn't need to.  There is now a member `_specs_already_associated` which tracks the index hashes for which we've already put the specs into the `_mirrors_for_spec`.  If an index hash is in that dictionary, there should be no need to read the index, iterate the specs, etc, so I avoid it in in that case.  At this point it was my opinion that `rengenerate_spec_cache()` does almost doing nothing if it is called multiple times in succession.  Do you not agree with that assessment?

scheibelp replied:
That will work although I think it would be more straightforward if that data structure was a set (and you check if `cached_index_hash in already_associated` vs. `already_associated[cached_index_hash] == True`).

I think I would still have a slight preference for basing this on whether `_mirrors_for_specs` was populated at all, since I think the `regenerate_spec_cache` function will always do the work of populating it (so in some sense it feels slightly redundant to have an additional data structure to track this state).

scheibelp commented on lib/spack/spack/binary_distribution.py:
```diff
@@ -852,117 +1216,98 @@ def extract_tarball(spec, filename, allow_root=False, unsigned=False,
+    nothing has changed on the configured mirrors.)
+    """
+    cache_manager.update_local_index_cache()
+    cache_manager.regenerate_spec_cache()
```
This works but can you make it so that `update_local_index_cache()` calls `regenerate_spec_cache()`? I think that users would always want to call these together, so we should combine them.

scheibelp reviewed (changes requested):
I have more requests all related to `BInaryDistributionCache`. Most are just renaming functions but there are a couple slightly-more-involved requests. Let me know if you think any are not worthwhile.

scottwittenburg committed:
More changed names, descriptive comments, and a couple clarifying changes

scottwittenburg commented:
@scheibelp I think I've addressed all the things mentioned in your latest pass, as well as the stuff we discussed earlier today.  The one thing that I kept, contrary to what we talked about, was that in `regenerate_spec_cache`, I did not simply use the truthy value of `_mirrors_for_spec` as the guard against doing all the processing again.  My reasoning for that was that if we got there because a new mirror was added since the last `udpate()` (and there were no other changes), then we shouldn't need to ingest the specs from all the cached indices again, only the one from the new mirror.  However I did change `_specs_already_associated` to a `set` as we discussed.

scottwittenburg commented:
Maybe I shouldn't have rebase on the latest `develop`.  While this change didn't actually touch it, `crypto.py` is now failing the vermin check:

```
!2, 3.9      /Users/scott/projects/spack/lib/spack/spack/util/crypto.py
  L7 C7: 'hashlib' module requires 2.5, 3.0
  L24 C16: 'dict' member requires 2.2, 3.0
  L44: 'str.format' member requires 2.6, 3.0
  L48: 'hashlib.new(usedforsecurity)' requires !2, 3.9
  L95: `with` requires 2.5, 3.0
  L134: function decorators requires 2.4, 3.0
  L150: 'sys.version_info' member requires 2.0, 3.0
  L157 C16: 'enumerate' member requires 2.
```

scottwittenburg commented:
Although actually, a new version of `vermin` was released 34 minutes ago.

**edit** And I just verified that with the previous version (`vermin==0.10.5`), this PR passes that check, but with the version released in the last hour (`vermin==1.0.0`) it hits the problem above.

scottwittenburg commented:
Did someone manually retrigger the `linux tests / unittests (3.9)` on this PR?  It had failed in a weird and inexplicable way to do with the gpg agent, but now it's running again.  Or is there maybe there auto-retry?

scheibelp commented:
> Did someone manually retrigger the linux tests / unittests (3.9) on this PR? 

Yes - me (sorry). The issue looked transient so I restarted the test

Also I'm looking at the vermin issue - I'll have to push a separate PR that addresses it and I'll have to rerun the vermin check here once it is merged

scheibelp commented:
See: https://github.com/spack/spack/pull/19612

scottwittenburg commented:
> > Did someone manually retrigger the linux tests / unittests (3.9) on this PR?
> 
> Yes - me (sorry). The issue looked transient so I restarted the test

Ah, no worries, thanks for doing that.

> Also I'm looking at the vermin issue - I'll have to push a separate PR that addresses it and I'll have to rerun the vermin check here once it is merged

Ok, sounds good, thanks for this too 😄

scheibelp commented:
Close/reopen to use #19612

scottwittenburg commented:
The gitlab pipeline failed at the `automake` job.  I think it's due to a bug in gitlab that it didn't report status, I've seen this before when some jobs couldn't run due to failure of a dependency.  I'm looking into the automake failure, but I feel like I've seen some issues mentioned in slack this week, so maybe there's already a PR to fix it.

scottwittenburg commented:
Can I remove the hash-change label?  I don't think it's true anymore.  Or does it not matter?

scheibelp commented:
> Can I remove the hash-change label? I don't think it's true anymore. Or does it not matter?

Yes - I did it: I agree the PR doesn't do that anymore and the tags should reflect it

scheibelp commented on lib/spack/spack/schema/config.py:
```diff
@@ -87,6 +87,7 @@
                 ],
             },
             'allow_sgid': {'type': 'boolean'},
+            'binary_distribution_cache_root': {'type': 'string'},
```
Sorry one last thing: I think this needs to be `binary_index_root` (I requested a name change in `binary_distribution.py` which was made there for `config:binary_index_root`).

scottwittenburg replied:
Dang it!  Should be just a sec.

scottwittenburg replied:
I'm sorry I missed that, but thank you for catching it.  I just pushed the fix.

scottwittenburg committed:
Fix overlooked config name change in schema

aweits commented:
I've had to revert this in one of my working branches/trees - I have a version of mpfr installed from before def1d5e637 (one of the multiple times the checksum on the 4.0.2 patch changed). With this commit in place, new install/uninstall operations fail with this message:
```
==> Error: Couldn't find patch for package builtin.mpfr with sha256: f2d2a530acb5e70e1a9d5b80881dbb4a504d56535c4bc103d83e0bb630172029
```
I haven't had time to look at finding the deeper issue (yet)....

scheibelp commented:
> I've had to revert this in one of my working branches/trees - I have a version of mpfr installed from before def1d5e (one of the multiple times the checksum on the 4.0.2 patch changed). With this commit in place, new install/uninstall operations fail with this message:

There is an issue at https://github.com/spack/spack/issues/19649 and a fix-in-progress (should be merged in a few hours) at https://github.com/spack/spack/pull/19672
